src/firmware_new/boards/frankie_v1/udp_multicast.c

```diff
@@ -0,0 +1,117 @@
+#include "firmware_new/proto/control_fw.pb.h"
+#include "lwip/api.h"
+#include "lwip/inet.h"
+#include "lwip/ip_addr.h"
+#include "lwip/memp.h"
+#include "lwip/opt.h"
+#include "lwip/pbuf.h"
+#include "lwip/sys.h"
+#include "lwip/tcp.h"
+#include "lwip/udp.h"
+#include "main.h"
+#include "pb.h"
+#include "pb_decode.h"
+#include "pb_encode.h"
+
+// TODO the messages are global for now, will need to be moved
+// when this file is properly integrated. This file only serves
+// as a reference implementation of reliable multicast with IPv6
+robot_ack ack       = robot_ack_init_zero;
+control_msg control = control_msg_init_zero;
+
+// this buffer is used to hold serialized proto. Allocated
+// on the data segment to avoid mallocing on every packet received
+static uint8_t buffer[robot_ack_size];
+
+// the thread that handles multicast recv/send
+static void udp_multicast_thread(void *arg);
+
+struct multicast_config
+{
+    ip_addr_t multicast_address;
+    unsigned multicast_port;
+    unsigned send_port;
+} typedef multicast_config_t;
+
+/*
+ * Thread that creates a send and recv socket, joins the specified
+ * multicast group and start listening for packets. Increments a counter
+ * on every multicast packet and sends a unicast packet back to the sender.
+ *
+ */
+static void udp_multicast_thread(void *arg)
+{
+    multicast_config_t *config = (multicast_config_t *)arg;
+    struct netbuf *rx_buf, *tx_buf;
+    int msg_count = 0;
+
+    // TODO proper err handling, for now all errs are ignored
+    err_t err;
+
+    // create two new UDP connections on the heap
+    // NOTE: we need two seperate UDP sockets as we will be
+    // receiving multicast but sending unicast packets to avoid
+    // network congestion (peer robots don't need to know the status
+    // of other robots)
+    struct netconn *recvconn = netconn_new(NETCONN_UDP_IPV6);
+    struct netconn *sendconn = netconn_new(NETCONN_UDP_IPV6);
+
+    // bind the socket to the multicast address and port
+    // we then use that connection profile to join the specified
+    // multicast group. For the send socket, we use IP6_ADDR_ANY
+    // which will default to the ETH interface.
```

Comment:
nit: Comment does not describe how the send socket will work if the ETH interface is unavailable / we aren't using ETH. Or is this currently not set up to send back over the network for now?

Reply:
we will always be using ETH so we can always assume it exists? there is no other way to send data. 

It is setup to send back on the network

Reply:
Oh right, ETH can be redirected to wifi because of big-brain RMII switching. I just saw ETH and thought of the actual ethernet connection. _Maybe_ worth a clarifying comment but no strong feelings. It pretty core to our design so probably doesn't need to be commented _everywhere_.

Reply:
changed